### ISSUE TITLE
Towards a strongly-typed taichi

### ISSUE THREAD

**Issue by re-xyr** [taichi-dev/taichi#4145]
Taichi is *statically typed*, but not strongly typed. What that means is that taichi has implicit conversion. There are many arguments against implicit conversion:

- It encourages users to ignore the difference between different datatypes, creating higher risks of precision loss and over/underflow.
- Maintaining a set of implicit conversion rules is tedious and error-prone.
- Users are often not exactly sure where the conversion will be inserted, so they can't reliably predict program behavior.
- Only "builtin" operators have implicit conversion; other internal/external functions, or (in the future) user-defined functions do not. This creates a distinction between operators and other functions, which complicates the language's AST structure *and* type system. Should there be no implicit conversions, operators and other functions can share a common implementation.
-  Relying on conversion too much is a code smell and suggests refactoring or rethinking the program structure. Allowing implicit conversion is obfuscating that.

Therefore I propose that taichi moves towards strong typing, i.e. removal of implicit conversion. This is a breaking change so it can happen in conjunction with v1.0 release.

*Apparently there may also be many reasons against this change and I'm happy to hear them.

**Comment by lin-hitonami** [comment]
Implicit type conversion is a useful feature and many programmers like it. Maybe we will lose a portion of users if we do not have it. However, I think we can make implicit conversion a warning, and add an option to treat specified warning as error.

**Comment by k-ye** [comment]
I also agree that we should strike a balance between ease-of-use and type system rigor. That said, we should definitely keep it to a minimum of where implicit conversions are allowed to happen. We should consider identifying all kinds of implicit conversions  in Taichi now, define the allowed cases, and bake them into our spec. Thoughts? @strongoier

**Comment by re-xyr** [comment]
Basically the `BinaryOp` casting now happens like this:

- If the operation is *bitwise* and either side is not *integral*, fail.
- If the operation is *true division* and either side is not *real*, cast it to *default float type*.
- If the type of two sides are not the same,
  - If any side is *custom type*, cast it to its *compute type*.
  - Now resolve the *result type* via [C++ implicit cast rules](https://en.cppreference.com/w/cpp/language/operator_arithmetic#Conversions). Cast both sides to the *result type*.
    - The implementation of this is rather... crude. There's apparently one thing called `std::common_type<T1, T2>::type` while we're using the brute-force `decltype(std::declval<T1>() + std::declval<T2>())`. 

Other casting behaviors include:
- Casting value type to destination type in `StoreStmt` and `AtomicOp`s
- Casting indices to `i32`s - reasonable, but we should check if they are integral in the first place
- `UnaryOp`s, i.e. `round`, `floor`, `ceil` and trig functions will fail if the operand is not *integral* - this is surprisingly stricter than `BinaryOp`s

**Comment by strongoier** [comment]
IMO how to make the tradeoff should really depend on who is the target audience of Taichi. I'd suggest evaluating this in more detail, e.g., measuring how much existing Taichi code (examples might be a good start) will break if we cancel implicit conversion, and making a survey of Taichi users about whether they can accept this or not.

**Comment by re-xyr** [comment]
Because *my* end goal is to unify the underlying implementation of internal functions and operators, and make them all properly typechecked, another way of achieving that is to figure out a way to express casting behaviors in the type system like this:

```
add<L : Primitive, R : Primitive>(lhs: L, rhs: R) : upper_bound(L, R)
```

**Comment by ailzhang** [comment]
I like this proposal but we might not be able to get to it as people said above. 
A really easy experiment is just remove one of the implicit cast we have right now and see how things work. My guess is that might already indicate the objection from users. 
A smaller step forward is that we can document our implicit casting behavior well for now, and evaluate whether to keep / remove them one by one.

**Comment by RuRo** [comment]
Not sure, if this is the correct issue to comment, but I honestly don't understand the motivation for the current integer casting rules.

For example, in C++ `decltype(std::declval<uint8_t>() + std::declval<uint8_t>())` is `int32_t`. This makes sense because it means that the computation is performed "as if" in a wide integer type and the user can then do stuff like this:
```cpp
uint8_t a = 200, b = 100;
int32_t wide_integer = b - a; // -100, no over/underflow
```
This is called [integral promotion](https://en.cppreference.com/w/cpp/language/implicit_conversion#Integral_promotion).

However, there is an important caveat for this rule to make sense.
Assuming two's complement signed integers or any unsigned integers,
```
uint8_t -> int32_t -> (do computation) -> int32_t -> uint8_t
```
is indistinguishable from
```
uint8_t -> (do computation) -> uint8_t
```
and so the following computation
```cpp
uint8_t modulo_arithmetic = b - a;
```

1) Is 100% valid and doesn't generate any warnings even with `-Wconversions`
    (compare with)
    ```cpp
    uint8_t downcasting = wide_integer;
    // warning: conversion from ‘int32_t’ {aka ‘int’} to ‘uint8_t’ {aka ‘unsigned char’} may change value [-Wconversion]
    ```
    
2) The compiled code doesn't actually need to promote `a` and `b` to 32-bit values if it can perform the computation using 8-bit arithmetic (and if it is faster to do so)

Taichi currently spams warnings about precision loss at almost every `u8` operation:
```python
import taichi as ti
ti.init()
@ti.kernel
def foo(arg: ti.u8):
    temp = arg
    # or
    temp: ti.u8 = arg
    # or even
    temp: ti.u8 = ti.cast(arg, ti.u8)
foo(0)
```
Results in
```
[W 03/14/22 20:08:44.069 938360] [type_check.cpp:visit@142] [$2] Local store may lose precision (target = u8, value = u8), at
2
[W 03/14/22 20:08:44.069 938360] [type_check.cpp:visit@142] [$4] Local store may lose precision (target = u8, value = u8), at
4
[W 03/14/22 20:08:44.069 938360] [type_check.cpp:visit@142] [$7] Local store may lose precision (target = u8, value = u8), at
7
```

Am I missing something? Is there any way to silence these warnings?

**Comment by strongoier** [comment]
@RuRo I think it is a bug in Taichi. For these "lose precision" warnings, Taichi compares the destination type with the common type of the destination type and the source type: https://github.com/taichi-dev/taichi/blob/ceb4aa683ef7b7f17100b54665628da94afaa24c/taichi/transforms/type_check.cpp#L140

As the common type (for arithmetic expressions) of `u8` and `u8` is `u32`, Taichi considers it as a down-casting, which does not make sense.

**Comment by RuRo** [comment]
Yes. And also it should never warn about precision loss when it is caused by an explicit `ti.cast` IMHO. Nobody in their right mind would do `ti.cast(arg, ti.u8)` and be surprised that the value of `arg` is truncated.
